File: frontend/types/order.ts
// Khớp CHÍNH XÁC `OrderRead`/`OrderItemRead`/`OrderCreate` thật
// (backend/app/schemas/order.py, task 3.4.2 + shipping_name thêm ở task
// 4.3.2) - bản trước đây (camelCase, "completed" thay vì "delivered", thiếu
// shipping_*/note) là placeholder lệch hẳn, đã bị flag ở docs/KNOWN_TODOS.md #20.

export type OrderStatus = "pending" | "confirmed" | "shipping" | "delivered" | "cancelled";

export interface OrderItem {
  id: number;
  product_id: number;
  product_name: string;
  quantity: number;
  // string, KHÔNG PHẢI number - cùng lý do Product.price/CartItem.unit_price:
  // Pydantic serialize Decimal thành string.
  price_at_purchase: string;
}

export interface Order {
  id: number;
  user_id: number;
  status: OrderStatus;
  total_amount: string;
  shipping_name: string;
  shipping_address: string;
  shipping_phone: string;
  note: string | null;
  items: OrderItem[];
  created_at: string;
  updated_at: string;
}

// Payload POST /orders - snapshot lúc đặt, KHÔNG có payment_method (hệ thống
// hiện chỉ hỗ trợ COD, xem CLAUDE.md mục Notes).
export interface OrderCreatePayload {
  shipping_name: string;
  shipping_address: string;
  shipping_phone: string;
  note?: string;
}

File: backend/app/schemas/order.py
"""Pydantic schemas: Order (task 3.4.2) - viết lại hoàn toàn từ placeholder
cũ, khớp model `Order`/`OrderItem` thật (task 3.1.3)."""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.order import OrderStatus
from app.schemas.base import BaseSchema


class OrderCreate(BaseModel):
    """`shipping_address`/`shipping_phone` BẮT BUỘC (khớp NOT NULL ở model) -
    CỐ TÌNH không tự điền từ `users.address`/`users.phone` phía server: đúng
    thiết kế "snapshot lúc đặt" đã ghi ở docs/DATABASE_SCHEMA.md (mục
    `orders`) - giá trị PHẢI đến từ chính request checkout (client tự
    prefill từ profile ở UI nếu muốn, đó là việc của Frontend, không phải
    Backend ngầm suy ra)."""

    shipping_name: str = Field(min_length=1, max_length=150)
    shipping_address: str = Field(min_length=1, max_length=500)
    shipping_phone: str = Field(min_length=1, max_length=20)
    note: str | None = Field(default=None, max_length=500)


class OrderItemRead(BaseSchema):
    """Snapshot BẤT BIẾN tại thời điểm mua - CỐ TÌNH KHÔNG join dữ liệu sản
    phẩm hiện tại (giá/tên sản phẩm có thể đã đổi từ lúc mua) - đúng thiết kế
    `order_items` (xem `app/models/order.py`)."""

    id: int
    product_id: int
    product_name: str
    quantity: int
    price_at_purchase: Decimal


class OrderRead(BaseSchema):
    id: int
    user_id: int
    status: OrderStatus
    total_amount: Decimal
    shipping_name: str
    shipping_address: str
    shipping_phone: str
    note: str | None = None
    items: list[OrderItemRead] = []
    created_at: datetime
    updated_at: datetime


class OrderStatusUpdate(BaseModel):
    status: OrderStatus


class OrderListFilter(BaseModel):
    """Query param cho `GET /orders/admin` (task 3.4.2) - filter theo trạng
    thái/khoảng ngày đặt hàng, theo đúng mô tả docs/API_SPEC.md mục 5."""

    status: OrderStatus | None = None
    date_from: date | None = None
    date_to: date | None = None
